### DISCUSSION TITLE
some newbie questions about JoltPhysics

### DISCUSSION THREAD

**Discussion by brightening-eyes** [jrouwe/JoltPhysics#410] [Q&A] [answered]
Hello,
I have some questions which might seem fullish
* when moving my character, can I use step listeners to check for user input and change the velocity accordingly? (I saw vehicle constraint's code and it was a step listener).
* how to pair a character with my virtual character (in order for other objects to see it)? I want to check collision with vehicles and other objects.
* in character's API, I didn't see AddBuoyancyImpulse in order to make it floating in the water.
* for my NPCS, should I use character or CharacterVirtual?
* how to have Air resistance when my character falls
* and, how to use vehicles to create aircrafts? (simulating stalls and so on is what I want to do as well).

**Comment by jrouwe** [comment, marked as answer]
> * when moving my character, can I use step listeners to check for user input and change the velocity accordingly? (I saw vehicle constraint's code and it was a step listener).

You could, but usually you would move (update) your character in your game code, so you can update the input then as well.

> * how to pair a character with my virtual character (in order for other objects to see it)? I want to check collision with vehicles and other objects.

Please take a look at this thread: https://github.com/jrouwe/JoltPhysics/discussions/239

> * in character's API, I didn't see AddBuoyancyImpulse in order to make it floating in the water.

That API doesn't exist. Usually the player behaves in a completely non-physical way, so I didn't bother creating it. I would create something simple but you can also take a look at that function (or 'Game Programming Gems 6' section 2.5 Exact Buoyancy for Polyhedra) for inspiration.

> * for my NPCS, should I use character or CharacterVirtual?

It depends on how your NPCs navigate. If they already know where they can walk by means of e.g. a navmesh you can get away with a Character (and one that only collides with dynamic objects because there's no point in colliding with static geometry if you already know where you can walk due to your navmesh). That class is a lot cheaper. If you need things like stair walking for your NPCs you're going to have to use the much more expensive CharacterVirtual.

> * how to have Air resistance when my character falls

See e.g. here: https://www.geeksforgeeks.org/air-resistance-formula/

> * and, how to use vehicles to create aircrafts? (simulating stalls and so on is what I want to do as well).

The VehicleConstraint is made specifically for ground vehicles, I don't think it would make sense to create an air vehicle out of it (although you might use it for the landing gear). I would start by creating a regular Body and use AddForce/Impulse to control it, but I'm not an expert on air plane physics, you have to do some Googling yourself.
